hazelcast/src/main/java/com/hazelcast/jet/impl/execution/OutboxImpl.java

```diff
@@ -156,15 +159,20 @@ private boolean offerInternal(@Nonnull int[] ordinals, @Nonnull Object item) {
             unfinishedItem = null;
             unfinishedItemOrdinals = null;
             if (item instanceof Watermark) {
-                long wmTimestamp = ((Watermark) item).timestamp();
+                Watermark wm = (Watermark) item;
+                long wmTimestamp = wm.timestamp();
                 if (wmTimestamp != WatermarkCoalescer.IDLE_MESSAGE.timestamp()) {
                     // We allow equal timestamp here, even though the WMs should be increasing.
                     // But we don't track WMs per ordinal and the same WM can be offered to different
                     // ordinals in different calls. Theoretically a completely different WM could be
                     // emitted to each ordinal, but we don't do that currently.
-                    assert lastForwardedWm.get() <= wmTimestamp
-                            : "current=" + lastForwardedWm.get() + ", new=" + wmTimestamp;
-                    lastForwardedWm.set(wmTimestamp);
+                    if (lastForwardedWm.get(wm.key()) == null) {
+                        lastForwardedWm.put(wm.key(), SwCounter.newSwCounter(Long.MIN_VALUE));
+                        return true;
```

Comment:
Why do we return here?

We can also use `computeIfAbsent` instead.

Reply:
I used `putIfAbsent` + assert + `put`.